### DISCUSSION TITLE
DSL for Customization

### DISCUSSION THREAD

**Discussion by quark-zju** [markbt/streampager#47] [Ideas]
There are users who wants less because they dislike the ruler color choices (https://fburl.com/3yug85rs). And there seem to be need for customized key bindings too. I wonder if we can have some DSL and a mini interpreter to provide flexibility for those cases.

For the DSL, I'm thinking about utilizing serde to do the parsing (to AST) work so we don't need to reinvent the language and avoid pest/buck issues. serde_json might be good enough. [serde_lexpr](https://docs.rs/serde-lexpr/0.1.2/serde_lexpr/) might make it slightly more friendly. Different places define different local symbols.

For example, the ruler might be written as:

```
(colorize white black
 (left-right-items title (list repeat-count file-info position loading)))
```

where `repeat-count`, `file-info` etc. are local symbols only available to the ruler context.

```
(match-key (ctrl a) (action-x))
```

If we allow local states, and expose stuff like the current line number, then we might be able to implement the marker feature (`m a`) entirely from the DSL. If we have `cond`, then we might be able to handle same key differently based on state (ex. Escape based on "repeat count").

Not trying to make this evolve into emacs, but it seems the amount of code needed to support such flexibility is not a lot and it can probably satisfy a large range of customization needs.

Let me know if you think this is a good idea and whether you want to implement it or I should go ahead write something.

**Comment by antonok-edm** [comment]
@quark-zju I think keymappings are already fully implemented by now. I opened https://github.com/markbt/streampager/issues/56#issue-1260546129 for further tracking bar style configuration.
